### ISSUE TITLE
Mixing :is() (or equivalent) with pseudo-elements

### ISSUE THREAD

**Issue by tabatkins** [w3c/csswg-drafts#9702]
The original version of `:is()`, `:-webkit-matches()`, allowed pseudo-elements in its argument list. So you could write `.foo:-webkit-matches(*, ::before, ::after)`, and it would be equivalent to `.foo, .foo::before, .foo::after`.

We ended up removing this, because it conflicts with the Selectors data model - simple selectors, like pseudo-classes, only ever *filter* the current set of matched elements; only combinators can *change* the set to new elements (and pseudo-elements). The pseudo-element "selector" is a legacy syntax mistake from the early days of CSS, and today would have been instead done as a type of combinator (as discussed in #7346, which still might happen).

But this means that you can't easily write useful selectors like the above. You *can* write them with Nesting, but you can't then nest *further*: `.foo { *, &::before, &::after {...}}` is fine, but `.foo, .bar::before { &:hover {...}}` won't match any hovered ::before pseudos, despite `.bar::before:hover` being a valid and meaningful selectors, because it's treated identically to `:is(.foo, .bar::before):hover`, which drops the `.bar::before` arg from the `:is()`.

This isn't great! And it makes Nesting less useful for future cases of nested pseudo-elements, in addition to anything built on Nesting concepts, like Mixins probably will be.

I don't have an answer ready for this, I just needed to raise it for broader discussion and thought.

---------

Some loosely-ordered thoughts:

* The three syntactic elements of Selectors are simple selectors (filter the matched elements), combinators (change the matched elements), and *lists* (union the matched elements).
    * Possibly we recognize a fourth - the "complex selector unit" which filters *and* might change the subject to a pseudo-element.
* `:is()` is trying to replicate the *functionality* of lists, but pulling it into the syntactic space of simple selectors, thus the conflict. You can combine `:is()` with other simple selectors, so the final subject of each selector inside of `:is()` needs to be the same element.
* If we did allow pseudo-elements in there *and* kept the data model correct, it would be unintuitive; we'd effectively have some selector reordering. That is, `:hover:is(*, ::before)` would have to be equivalent to `:hover, ::before:hover`, not `:hover, :hover::before`.
* But that sort of re-ordering is in conflict with things like `.foo:is(::before, ::after)`, wanting to avoid repeating the preceding selector. And simple selector order can't matter; `.foo:is()` and `:is().foo` have to be identical. So that also prevents reasonable stuff like `.foo:is(::before, ::after):hover` from working.
* I feel like we need to somehow invent a way to wedge "lists" into the syntax of a single selector, rather than being restricted to the top level of a selector as it is now. The problem isn't the branching that :is() allows, it's the fact that :is() is a simple selector, so we can't care about order/etc.
* Spitballing: naked parens indicate a list. Lists can change the subject, and stuff can't get reordered across them. Like `.foo(*, ::before, ::after):hover`.
* Nope, `foo(` is a function token. I need the parens to be separate from other bits of the compound selector, but I also need to still be able to glom a compound selector before or after it.
* Second try: `.foo = (*, ::before, ::after) = :hover`. 
    * `=` is a new combinator that *doesn't* change the set of matched elements at all; it's a no-op. 
    * `(...)` is a new term in the `<complex-selector-unit>` grammar, that takes a selector list and matches any of them. Since it's part of a complex selector and *guaranteed* to be surrounded by combinators, we avoid the issues from earlier.
    * So in this example, we (1) select `.foo` elements, (2) without changing the set of matched elements, select `*, ::before, ::after` elements (aka the element, plus their before/after pseudos), (3) without changing the set of matched elements, select `:hover` elements.
    * This works with other combinators, too: `.foo > (*, ::before)` selects all children of a `.foo` element *and* their before pseudos.
    * The selectors inside the `()` can be full complex selectors, just like in `:is()`, so `.foo = (.bar *, #baz > *)` selects `.foo` elements that have a `.bar` ancestor or a `#baz` parent.
    * This is still slightly weaker than Nesting itself; you can't do the equivalent of `.foo { &, & > .bar {...}}` because the current match set (whatever the preceding combinator yielded) is always the subject of the selectors in the paren list.  That's probably okay. We'd have to invent a new way to refer to the elements being matched if we wanted to expand that (not `:scope` or `&`, but a secret third way). I dunno, maybe we do need it.
    * Specificity would have to be the same as `:is()`, to avoid the same sort of "which way did it match" exponential explosion.
    * Nesting can then define itself on top of this - `X { Y {...}}` desugars to `(X) = (Y) {...}`, with the magic ability to use `&` to explicitly refer to the current match elements. If I'm thinking about this correctly, this is a no-op change *except* that it allows `::before, ::after { &:hover {...}}` to work, where today it doesn't.

**Comment by tabatkins** [comment]
Yeah, I'll give it some more time to percolate, but some variant on my "Second Try" idea is probably what I'm aiming for. Pulling it out to make it easier to find/reference:

Problem statement: we'd like the following to be valid:

```css
.foo {
	&, &::before, &::after {
		&:hover {
			color: blue;
}}}
```

...and have the obvious result, equivalent to:

```css
.foo:hover, .foo::before:hover, .foo::after::hover {
	color: blue;
}
```

Solution: introduce a new combinator, `=`, and a new `<complex-selector-unit>`, the parenthesized list.

1. `=` selector is a no-op combinator; it has no effect on the current set of matched elements. So `.foo = .bar` is the same as `.foo.bar`, `.foo = ::before` is the same as `.foo::before`, etc.
2. `<complex-selector-unit>` currently expressed the `compound selector + pseudo-element" bit of the grammar. It both filters the current matched elements *and* possibly changes the set (just to pseudos, currently), and might do that repeatedly. We add `(<selector-list>)` to the grammar for it.
	* We just run the selector list on the current set of matched elements, filtering it and potentially changing it (just to pseudos, currently) 
	* The problem statement above could be written as `.foo = (*, ::before, ::after) = :hover {...}`.
    * This works with other combinators, too: `.foo > (*, ::before)` selects all children of a `.foo` element *and* their before pseudos.
    * The selectors inside the `()` can be full complex selectors, just like in `:is()`, so `.foo = (.bar *, #baz > *)` selects `.foo` elements that have a `.bar` ancestor or a `#baz` parent.
    * This is still slightly weaker than Nesting itself; you can't do the equivalent of `.foo { &, & > .bar {...}}` because the current match set (whatever the preceding combinator yielded) is always the subject of the selectors in the paren list.  That's probably okay. We'd have to invent a new way to refer to the elements being matched if we wanted to expand that (not `:scope` or `&`, but a secret third way). I dunno, maybe we do need it.
    * Specificity would have to be the same as `:is()`, to avoid the same sort of "which way did it match" exponential explosion.

Nesting can then define itself on top of this: `X { Y {...}}` desugars to `(X) = (Y) {...}`, with just a little extra magic:
	* `&` refers to the current matched elements (so in the `(Y)` bit, `&` refers to the `(X)` elements, etc). (This expands lists' ability to change the set from just changing to pseudo-elements, to changing to anything potentially.)
	* Remember that `&` represents the elements (and now, pseudo-elements) matched by the parent selector; it's *not* textual substitution. So `::before { :hover& {...}}` is equivalent to `::before:hover`, not `:hover::before` - the parent selector filters the match set to all the ::before pseudos, then `:hover&` represents elements that are (a) in the parent's match set, and (b) hovered.
	* The final list keeps "normal" specificity behavior, where it matters which selector actually matched. (This avoids changing the *current* behavior of `.foo, { .bar, #baz {...}}`.)

**Comment by Loirooriol** [comment]
While this can be useful on its own, I don't see why nesting needs `=`? Wouldn't it suffice to change https://drafts.csswg.org/css-nesting/#nest-selector

> the [nesting selector](https://drafts.csswg.org/css-nesting/#nesting-selector) represents the elements matched by the parent rule

into

> the [nesting selector](https://drafts.csswg.org/css-nesting/#nesting-selector) represents the elements <ins>or pseudo-elements</ins> matched by the parent rule

**Comment by tabatkins** [comment]
No, because you still need to distinguish `::before:hover` from `:hover::before`, while `&:hover` and `:hover&` are meant to be identical in the Selectors data model.

**Comment by Loirooriol** [comment]
Well yeah, `&:hover` and `:hover&` would both behave like `::before:hover`

`:hover::before` could be done with #7346, like `:hover :>> &`.

`=` still doesn't seem required for nesting. Also, I think this character can be confusing because it seems to be aliasing selectors or something? I would prefer `%` or such.

**Comment by tabatkins** [comment]
> Also, I think this character can be confusing because it seems to be aliasing selectors or something?

I'm not sure what you mean by this. It's a combinator (just a no-op one).

**Comment by Loirooriol** [comment]
The `=` symbol is too tied to equality/assignment in my mind, so at 1st glance `.foo = .bar` seems to mean that elements with a `foo` class will also match styles for a `bar` class or something like that, I don't know. Maybe it's just getting used to it, but I would prefer representing this combinator with another symbol.

**Comment by yisibl** [comment]
Yes, we need this feature.

For example, if you want to set the input track for both browsers at the same time, you need to write the same rule set twice, and it won't work in `:is()`.

```css
input[type=range]:is(
    ::-moz-range-track,
    ::-webkit-slider-runnable-track
) {
    background: #ddd;
}
```

**Comment by Loirooriol** [comment]
@yisibl That should already be doable as

```css
input[type=range] {
  &::-moz-range-track, &::-webkit-slider-runnable-track {
    background: #ddd;
  }
}
```

Or #7346 proposes

```css
input[type=range] :> :is(-moz-range-track, -webkit-slider-runnable-track) {
    background: #ddd;
}
```

**Comment by yisibl** [comment]
@Loirooriol yes, the core point here is the need for a pseudo-element selector that supports [forgiving-selector-list](https://drafts.csswg.org/selectors-4/#typedef-forgiving-selector-list).

**Comment by jpzwarte** [comment]
So i kind of got redirect here :) I ran into the following issue:

This works:

```css
sl-month-view::part(finish) {
  background: var(--sl-color-success-plain);
  border-radius: 50%;
  color: var(--sl-color-text-inverted);
}

sl-month-view::part(finish):hover {
  background: var(--sl-color-success-bold);
}
```

But this doesn't:

```css
sl-month-view::part(finish) {
  background: var(--sl-color-success-plain);
  border-radius: 50%;
  color: var(--sl-color-text-inverted);

  &:hover {
    background: var(--sl-color-success-bold);
  }
}
```

This behavior is not obvious to me and it feels like both should work just fine (from a web dev point of view).

@tabatkins can you confirm i'm in the right issue for this?

**Comment by andruud** [comment]
> * Remember that `&` represents the elements (and now, pseudo-elements) matched by the parent selector; it's _not_ textual substitution. So `::before { :hover& {...}}` is equivalent to `::before:hover`, not `:hover::before` - the parent selector filters the match set to all the ::before pseudos, then `:hover&` represents elements that are (a) in the parent's match set, and (b) hovered.

Then ideally we would disallow `&` except at the start of the compound? It's not unreasonable to expect `:hover::before` here. I get that it's consistent with the "filtering" model, but this will surely be seen as your selector being silently re-ordered.

> you can't do the equivalent of `.foo { &, & > .bar {...}}`

This would then desugar to: `.foo = (&) { ... }, .foo = (& > .bar) { ... }`?

> `.foo = (*, ::before, ::after) = :hover {...}`

Are there any restrictions on where lists with pseudos can appear

- Can anything follow it (that is not allowed parse time today)? `.a = (.b, ::after) = .c {...}`.

> Since it's part of a complex selector and guaranteed to be surrounded by combinators

What does `&` refer to here? `.a = (.b, .c = (&)) = .d`

> The pseudo-element "selector" is a legacy syntax mistake from the early days of CSS, and today would have been instead done as a type of combinator

Is there anything that could be improved in this proposal if we _first_ fix that mistake (as an alternative way to access pseudo elements), and then make _that_ form work well with nesting?

**Comment by Loirooriol** [comment]
> Is there anything that could be improved in this proposal if we first fix that mistake

In https://github.com/w3c/csswg-drafts/issues/7346#issuecomment-1176857832 we already resolved on an (experimental) solution.

**Comment by tabatkins** [comment]
> Then ideally we would disallow & except at the start of the compound? 

Well, we don't currently have that limitation, and Nesting has been out for long enough that we probably can't change that.

> It's not unreasonable to expect :hover::before here. I get that it's consistent with the "filtering" model, but this will surely be seen as your selector being silently re-ordered.

I think it is somewhat unreasonable. For any *normal* element matched by the parent rule, `:hover&` in the child rule  will mean "the same elements as the parent, but only when they're hovered", same as `&:hover`. So `::before { :hover& {...}}` should match that - the same elements as the parent rule (some ::before pseudos), but only when they're hovered, aka `::before:hover`. It's no more re-ordering than `div { :hover& {...}}` is when it desugars to `div:hover`.

If you wanted the other way, you'd need the pseudo combinator, and  you'd write `::before { :hover :> & {...}}`.

> This would then desugar to: `.foo = (&) { ... }`, `.foo = (& > .bar) { ... }`?

No, since they're no longer nesting the `&` doesn't do the right thing. The (current) limitation in the proposal is that the `(<sel>)` can only change the subject via a pseudo-element selector; otherwise it's just filtering whatever the current unit would be otherwise. That's the issue with `&, & > .bar` - it's both retaining the parent's match set *and* moving to the match set's children element (and the filtering them with `.bar`). (But see the end of this comment for more exploration of this.)

> Are there any restrictions on where lists with pseudos can appear

No. It's just another option for the `<complex-selector-unit>` production, so you can chain after it with another combinator/etc

> Can anything follow it (that is not allowed parse time today)? `.a = (.b, ::after) = .c {...}`.

Yes, that's valid, and equivalent to `.a.b.c, .a::after.c {...}`.

> Is there anything that could be improved in this proposal if we first fix that mistake (as an alternative way to access pseudo elements), and then make that form work well with nesting?

No, that's not the side that's problematic for Nesting. Nesting has issues due to our decision to keep it as purely a sugar as possible, without causing combinatorial or exponential blowup in the desugaring. That means we're limited to what `:is()` can do, and `.foo:is(.bar, * :> before)` (using the `:>` "pseudo combinator" proposal we resolved on) also doesn't do what you're wanting. (It only matches ::before pseudos that are *also* `.foo` elements, aka `::before.foo`

The problem is that, unless we introduce a third way to refer to the "subject" coming from outside the `()` list, we're stuck with one of two options: either the subject is implicitly what the `()` selectors *match* (aka the last compound matches the subject) or it's implicitly what the `()` selectors *descend from* (aka the selectors are implicitly relative).

That is, we could move the selector *into* the parentheses, and then `.foo { &, & > .bar {...}}` could desugar to `.foo (= *, > .bar) {...}`, or we could leave the selector *outside* and then `.foo { &, .bar > & {...}}` desugars to `.foo = (*, .bar > *) {...}`. But we can't do both, and `:is()` already matches the second one, so staying consistent with that seems good.

The fact that `::foo` is *not* actually using the combinator machinery rescues us slightly and makes the case I'm trying to solve work - it just leans on the fact that pseudo-elements can already change the subject of the selector between combinators.

**Comment by andruud** [comment]
I am still trying to understand the ideas in this issue. I feel like this is more complicated than it needs to be, and/or that I'm missing something.

Looking at this another way for a bit, imagine that we have two (not necessarily web-exposed) primitives:

 * The pseudo-child combinator (`:>`), as [seen elsewhere](https://github.com/w3c/csswg-drafts/issues/7346#issuecomment-1176857832).
 * A pseudo-_class_, `:pseudo()`, which checks if some element has a specific pseudo element type. Example: `:pseudo(before)`.
   * This exists to make it clear that we're querying a pseudo element type (and not a tag name) _without_ having to care about adjacent combinators.

Looking at the original problem statement:

```
.foo {
  &, &::before, &::after {
    &:hover {
      color: blue;
    }
  }
}
```

We can think of that as the following:

```
.foo {
  &, & :> :pseudo(before), & :> :pseudo(after) {
    &:hover {
      color: blue;
    }
  }
}
```

Even if we use `:is()` directly, this should still match all of `.foo:hover, .foo::before:hover, .foo::after::hover` without violating the selectors data model:

```
:is(.foo, .foo :> :pseudo(before), .foo :> :pseudo(after)):hover {}
```

---

> I think it is somewhat unreasonable [that `::before { :hover& {} }` would desugar to `:hover::before { }`]

Thinking about it more, yeah, agree. Should be `::before:hover {}`.

**Comment by tabatkins** [comment]
> Even if we use `:is()` directly, this should still match all of `.foo:hover`, `.foo::before:hover`, `.foo::after::hover` without violating the selectors data model:

Unfortunately, it doesn't; reshuffling the syntax doesn't help us here, because the problem is much deeper in the Selectors data model.  When you have `:is(...)`, that is *implicitly equal* to `*:is(...)`. The `*` selector can only match elements that are currently visible in the tree, and that *does not* include pseudo-elements.

So, in `:is(.foo, .foo :> :pseudo(before), .foo :> :pseudo(after)):hover`, the subject of the `:is()` (the things matched against the `*`, or whatever else shared the compound with the `:is()`) are the `.foo`, `:pseudo(before)`, and `:pseudo(after)` parts, but those latter two can never match the `*` outside the `:is()`.

This is gonna be the problem for *anything* that works via a simple selector (like a pseudo-class) - the set of elements they are filtering is chosen before any of the filters are applied, and by default elements in "hidden" trees, like pseudo-elements, aren't going to be in that set.

That was the point behind my `=` combinator and the parenthesized syntax - the parenthesized syntax allows us to do something slightly different (because it's not part of a compound selector, there's nothing else limiting it to a particular set of matched elements), and the `A = B` combinator just says "elements that match A, and also match B".

-----------

Thinking on this a bit, I suppose another possibility would be to slightly change the Selectors data model, such that all the "hidden" elements *are* in the match set before each compound selector is applied, but they're *post-filtered* out of the match set if they weren't explicitly matched by something in the compound selector, before the next combinator is applied.

That is, `*` would select all elements *and pseudo-elements* on the page, but unless a `::foo` is in the `<complex-selector-unit>`, the pseudo-elements are removed from the final match result. That would allow `*:is(.foo, .foo::before)` to work (matching both `.foo` elements and their `::before` pseudos), while still preventing `.foo:is(*, ::before)` from working (it would only match `.foo` elements, as the `::before`s wouldn't match the `.foo` simple selector).

I wanna sit with this one a bit, but I think it might actually solve our problems generally, for both `:is()` and Nesting, without actually needing any new syntax at all?

**Comment by andruud** [comment]
> When you have `:is(...)`, that is implicitly equal to `*:is(...)`.

Thanks, that's the part I missed. It makes sense.

> Thinking on this a bit, I suppose another possibility would be to slightly change the Selectors data model, such that all the "hidden" elements _are_ in the match set before each compound selector is applied, but they're _post-filtered_ out of the match set if they weren't explicitly matched by something in the compound selector, before the next combinator is applied.

Yeah, I somehow thought it _already_ kind of worked like this.

> That is, `*` would select all elements _and pseudo-elements_ on the page, but unless a `::foo` is in the `<complex-selector-unit>`, the pseudo-elements are removed from the final match result. That would allow `*:is(.foo, .foo::before)` to work (matching both `.foo` elements and their `::before` pseudos), while still preventing `.foo:is(*, ::before)` from working (it would only match `.foo` elements, as the `::before`s wouldn't match the `.foo` simple selector).
> 
> I wanna sit with this one a bit, but I think it might actually solve our problems generally, for both `:is()` and Nesting, without actually needing any new syntax at all?

I think this is the way to go.

**Comment by tabatkins** [comment]
All right, Agenda+ to discuss this.

For clarity, the proposal here is that we slightly alter the Selectors data model, such that:

* *before* matching a `<complex-selector-unit>` (aka a compound selector + pseudo-elements, no combinators), we grow the set of currently-matched elements (which are about to be filtered by the compound) by adding all the pseudo-elements of all the currently-matched elements.
* *after* matching a compound selector, we filter out all the pseudo-elements that weren't matched explicitly by the `<complex-selector-unit>` (then apply the combinator and continue on, as normal)

This means that, for example, `:is(.foo, .foo::before, .foo::after):hover` will "work as expected":
1. We open, as usual, with the match set being the entire document.
2. We temporarily add all pseudo-elements of all elements to the match set.
3. We filter to only keep any elements/pseudo-elements that are hovered, and that are either `.foo`, `.foo::before`, or `.foo::after`.
4. We remove any remaining pseudo-elements that weren't explicitly matched (in this case, there's nothing to remove).
5. The end result is equivalent to `.foo:hover, .foo::before:hover, .foo::after:hover`.

However, a closely related selector, like `:is(*, ::before, ::after).foo:hover` will not match any pseudo-elements, because the `.foo` simple selector will filter out all the pseudo-elements.

For existing "normal" cases like `:hover`, we'll add all the pseudos, then remove all the pseudos (since none of them were mentioned in the complex-selector-unit), so there's no change in behavior.

For nesting, then, this will mean that the following will work:

```css
.foo {
  ...
  &, &::before, &::after {
    ...
    &:hover { ... } /* today, this is only equivalent to `.foo:hover`
  }
}
```

---------------

Some limitations to this approach:

1. As currently described, `.foo:is(*, ::before, ::after)` won't work. But this did work in the original `-webkit-matches()`. Is this limitation okay? If not, how can we adjust the model to allow that to work as well? Since simple selectors can be rearranged in a compound, this should presumably work identically to `:is(*, ::before, ::after).foo`, too. (If these two aren't identical, then this continues to be "not a simple selector" and needs more adjustments.)
2. As currently described, `:hover:is(::before)` and `:is(::before):hover` are identical - they're both equivalent to `::before:hover`. I think this is the more commonly intended ordering, but in cases where you want the other ordering, you can't express it with `:is()` or Nesting under this model. Is this limitation okay? (Assuming we have the pseudo-combinator, you could instead write `:hover :> :is(before, after)` or whatever, which would allow you to solve it when needed.)

**Comment by Crissov** [comment]
For what little it’s worth, my mental model does not like (naked) pseudo-elements nor the universal element selector inside `:is()`. Can we have a dedicated asterisk and colon-free pseudo-class function like `:as(element, before, after)` instead?

**Comment by Loirooriol** [comment]
Just to make sure, `::marker` and `*::marker` will still not match a marker originated by a before pseudo, right? Because the before is not explicitly matched.

**Comment by tabatkins** [comment]
@Crissov I'm not quite sure what you mean. Using asterisk inside of `:is()` is already allowed and functions normally, like `.foo:is(*, .bar, .baz)` (matching `.foo, .foo.bar, .foo.baz`). Could you elaborate on what's bothering you?

-------

@Loirooriol That's indeed my intention. `::before::marker` should be put in the match set for possible matching, but there will need to be some care in what "explicitly matched" means to prevent it from being matched by a plain `::marker`.

**Comment by Crissov** [comment]
Perhaps I just haven’t used `:is()` enough yet to see it. Your example makes little sense to me, because the selector matches the same as `.foo:is(*)`, except with higher specificity; that and forgiving list parsing then is the only difference from plain old `.foo`.

My imagination seems too limited to find a real use for `…:is(*, …)`.

**Comment by andruud** [comment]
> after matching a compound selector, we filter out all the pseudo-elements that weren't matched explicitly by the <complex-selector-unit> (then apply the combinator and continue on, as normal)

This means `:hover:is(::before)` can match, but `:is(:hover):is(::before)` can not, since the `:hover` compound within the `:is()` will have filtered away all pseudo-elements it didn't explicitly match. I'm not sure that sounds right?

EDIT: Formatting.

**Comment by tabatkins** [comment]
> I'm not sure that sounds right?

Right, I didn't intend for that. Maybe I'm meaning more along the lines of top-level compounds, so `:is()` arguments don't trigger the filtering? Unsure.

**Comment by bramus** [comment]
Isn’t the problem that `:is()` reads as the word _“is”_ in English, which is causing confusion here?

If you read it as _“is”_, the following (from the IRC log) indeed is confusing: `.foo:is(.bar, ::before, ::after) == .foo.bar, .foo::before, .foo::after` … there’s no way that a `.foo` _“is”_ also a `::before`.

But if you replace `:is()` with something like `:concat()`, it becomes more clear what is happening: `.foo:concat(.bar, ::before, ::after) == .foo.bar, .foo::before, .foo::after` … the pieces get glued together and that’s it.

---

Also, I tried to decipher the chat log but am not sure what exactly we resolved on for nesting. Does `.foo { &::before { &:hover {} } }` result in _hovering the `::before` of `.foo`_ or does it result _the `::before` when hovering `.foo`_? I think we resolved on the former (i.e. `.foo::before:hover`)?

**Comment by Loirooriol** [comment]
What I would expect, modulo specificity,

`.foo { &::before { &:hover{}}}` is `.foo::before:hover`
`.foo, ::before { &:hover {}}` is `.foo:hover, ::before:hover`
`.foo, ::before { :hover& {}}` is `.foo:hover, ::before:hover`
`.foo, ::before { &::marker {}}` is `.foo::marker, ::before::marker`
`.foo, ::before { ::marker& {}}` never matches
`.foo, ::marker { ::marker& {}}` is `::marker`

That is, `&` matches any element or pseudo-element matched by the parent selector.

And the problem with putting this into `:is()` seems that

- `::before:is(:hover)` should be `::before:hover` I think?
- `:hover { ::before& {}}` shouldn't match anything, because a top-level `:hover` can't match pseudo-elements.

**Comment by SebastianZ** [comment]
@bramus:

> Isn’t the problem that `:is()` reads as the word _“is”_ in English, which is causing confusion here?
> 
> If you read it as _“is”_, the following (from the IRC log) indeed is confusing: `.foo:is(.bar, ::before, ::after) == .foo.bar, .foo::before, .foo::after` … there’s no way that a `.foo` _“is”_ also a `::before`.
> 
> But if you replace `:is()` with something like `:concat()`, it becomes more clear what is happening: `.foo:concat(.bar, ::before, ::after) == .foo.bar, .foo::before, .foo::after` … the pieces get glued together and that’s it.

`:-webkit-matches()` was mentioned several times and @plinss also suggested to call it "matches" in the IRC log. So can't we just name it `:matches()` then?

`.foo:matches(.bar, ::before, ::after) == .foo.bar, .foo::before, .foo::after`

@jakearchibald also suggested to introduce pseudo-element combinators in #7346, which may help the mental model:

`.foo:matches(.bar, :> before, :> after)` == `.foo.bar, .foo::before, .foo::after`

And @Loirooriol suggested to introduce a new universal selector that includes tree-abiding pseudo-elements in #4565.

With that, `.foo { &**{ } }` == `.foo:matches(*, ::before, ::after, ::marker, ::before::marker, ::after::marker) { }` == `.foo, .foo::before, .foo::after, .foo::marker, .foo::before::marker, .foo::after::marker { }`.

Sebastian

**Comment by Loirooriol** [comment]
@SebastianZ While `**` would be allowed to match pseudo-elements, `&` refers to `.foo`, and pseudo-elements can't match `.foo`, so I think `.foo { &**{ } }` would just be like `.foo { }`

For such `:matches()`, I think it should then work with arbitrary combinators, e.g.

| Non-expanded | Expanded |
| - | - |
| `.foo:matches(.bar, .a > .b)` | `.foo.bar, .foo.a > .b` |
| `.foo:matches(.bar, .a::before)` | `.foo.bar, .foo.a::before` |
| `.foo:is(.bar, .a > .b)` | `.bar.foo, .a > .b.foo` |
| `.foo:is(.bar, .a::before)` | `.bar.foo, .a::before.foo` i.e. `.bar.foo` |

I don't know if this is how old `:webkit-matches()` behaved or not.

**Comment by tabatkins** [comment]
The `:is()` naming was a red herring, and has nothing to do with the pseudo-element discussion. Any confusion that might arise from that is independent of this topic; `:is(A, B)` was always meant to be read as "an element that **is** either A or B". I'll put that to the side; it's not germane to this issue.

-----------

Okay I just spent at least an hour working through some thoughts, then deleted it all. You're welcome. Here's the revised version of what I was working through, laying out the issues as clearly as I can. tldr: we have to solve a more general selector issue first in Nesting, before we can do anything elsewhere.

Ultimately, a pseudo-element is a combinator + some sort of simple selector. `A::before` is, spiritually, identical to `A > B`, it's just that the `::` "combinator" moves into the list of pseudo-children rather than the list of children that `>` moves into. So, any intuition and mechanics that apply to `>`, should also apply to pseudo-elements, and where there are holes we can see what might need to be patched.

If we look at nesting with `>` on the parent rule, the behavior of `A > B { &:hover {...}}` is clear - it's equivalent to `A > B:hover`. The same is true of `A > B { :hover& {...}}` - it's still equivalent to `A > B:hover`, because `&` is a simple selector and the relative order with `:hover` doesn't matter. The `&` matches the elements matched by the parent, which are B elements (that happen to be children of A elements); you intersect "the B elements" with "the :hover elements" and get your answer, identical in both cases. <sub>(But see below.)</sub>

Nobody really expects `A > B { :hover& {...}}` to be equivalent to `A:hover > B`, but if you do *want* that expansion instead, you can achieve it with a bit of repetition of the combinators: `A > B { :hover > & {...}}`.

Now take that to pseudo-elements. `A::before { &:hover {...}}` should be equivalent to `A::before:hover`; this is uncontroversial and what we resolved on at the meeting. Continuing the analogy, then, `A::before { :hover& {...}}` should *also* be equal to `A::before:hover`, because once again, it's matching `&` and `:hover` separately, then intersecting them: "the `::before` pseudoes (of A elements)" and "the :hover" elements.

If you do, for whatever reason, want `A:hover::before`, that's not currently achievable, because we lack the ability to reproduce the pseudo-element "combinator". However, we resolved to start experimental work on it in #7346 , spelling it `:>`. Using that, you could write `A::before { :hover :> & {...}}` and get the desired behavior, exactly paralleling the `>` case above.

-----------

Now nesting the other way around, with the `>` on the child rule: `:hover { & > B {...}}` and `:hover { A > & {...}}` are both straightforward, equivalent to `:hover > B` and `A > :hover`, respectively.

Moving to pseudo-elements, we'd naively write `:hover { &::before {...}}` and `:hover { ::before& {...}}`. I think this is probably still the correct thing to do, making them equivalent to `:hover::before` and `::before:hover`, respectively. We already distinguish between simple selectors before and after a pseudo-element, so that tracks here.

------

But now we come to the issue Oriol mentioned in [their previous comment last week](https://github.com/w3c/csswg-drafts/issues/9702#issuecomment-3533065434). In both `A::before { &:hover {...}}` and `:hover { ::before& {...}}`, the elements matching `:hover` are *by definition* not pseudo-elements, so you can't actually intersect them with the pseudo-elements (whether hidden in an `&` or stated directly as `::before`); you always get an empty intersection.

The issue here is that `::` *is* a weird "combinator". Unlike `A > B`, where the A and B elements live in the same "tree", in `A::before` the A and "before" live in different trees, and aren't visible at the same time. So a lot of simplifying assumptions we can make in the `A > B` case aren't true anymore. :(

And that's where I'm stuck so far. I need to put more thought into how to solve *this* issue, and if we can fix it, then I think the Nesting behavior falls right out.

And *then* we can talk about how to linearize a Nesting stack into a single selector, so you can actually write these sorts of selectors in querySelector()/etc if you want.

**Comment by romainmenke** [comment]
Been thinking about this one and I wonder if a [functional pseudo element selector](https://github.com/w3c/csswg-drafts/issues/13283) might help. This leans into the weirdness of `::` being a combinator.
